File: php/Deck.php
<?php

    abstract class Suit
    {
        const Spades = 0;
        const Hearts = 1;
        const Diamonds = 2;
        const Clubs = 3;
    }
    abstract class Rank
    {
        const Two = 2;
        const Three = 3;
        const Four = 4;
        const Five = 5;
        const Six = 6;
        const Seven = 7;
        const Eight = 8;
        const Nine = 9;
        const Ten = 10;
        const Jack = 11;
        const Queen = 12;
        const King = 13;
        const Ace = 14;
    }

    class Card
    {
        public $rank;
        public $suit;

        public function __construct($rank, $suit)
        {
            $this->rank = $rank;
            $this->suit = $suit;
        }
        public function toString()
        {
            $toReturn = "";
            switch($this->rank)
            {
                case 2:
                    $toReturn .= "Two";
                    break;
                    case 3:
                    $toReturn .= "Three";
                    break;
                    case 4:
                    $toReturn .= "Four";
                    break;
                    case 5:
                    $toReturn .= "Five";
                    break;
                    case 6:
                    $toReturn .= "Six";
                    break;
                    case 7:
                    $toReturn .= "Seven";
                    break;
                    case 8:
                    $toReturn .= "Eight";
                    break;
                    case 9:
                    $toReturn .= "Nine";
                    break;
                    case 10:
                    $toReturn .= "Ten";
                    break;
                    case 11:
                    $toReturn .= "Jack";
                    break;
                    case 12:
                    $toReturn .= "Queen";
                    break;
                    case 13:
                    $toReturn .= "King";
                    break;
                    case 14:
                    $toReturn .= "Ace";
                    break;
            }
            switch ($this->suit)
            {
                case 0:
                    $toReturn .= " of Spades";
                    break;
                case 1:
                    $toReturn .= " of Hearts";
                    break;
                case 2:
                    $toReturn .= " of Diamonds";
                    break;
                case 3:
                    $toReturn .= " of Clubs";
                    break;
            }
            return $toReturn;
        }
        public function toShortString()
        {
            $toReturn = "";
            switch($this->rank)
            {
                case 2:
                    $toReturn .= "2";
                    break;
                    case 3:
                    $toReturn .= "3";
                    break;
                    case 4:
                    $toReturn .= "4";
                    break;
                    case 5:
                    $toReturn .= "5";
                    break;
                    case 6:
                    $toReturn .= "6";
                    break;
                    case 7:
                    $toReturn .= "7";
                    break;
                    case 8:
                    $toReturn .= "8";
                    break;
                    case 9:
                    $toReturn .= "9";
                    break;
                    case 10:
                    $toReturn .= "10";
                    break;
                    case 11:
                    $toReturn .= "J";
                    break;
                    case 12:
                    $toReturn .= "Q";
                    break;
                    case 13:
                    $toReturn .= "K";
                    break;
                    case 14:
                    $toReturn .= "A";
                    break;
            }
            switch ($this->suit)
            {
                case 0:
                    $toReturn .= "s";
                    break;
                case 1:
                    $toReturn .= "h";
                    break;
                case 2:
                    $toReturn .= "d";
                    break;
                case 3:
                    $toReturn .= "c";
                    break;
            }
            return $toReturn;
        }
    }

    class Deck
    {
        public $_deck;

        public function __construct()
        {
            $this->_deck = array();
            for($i = 0; $i<=3; ++$i) {
                for($j = 2; $j<=14; ++$j)
                {
                    $this->_deck[] = new Card($j, $i);
                }
            }
        }
        public function shuffle()
        {
            shuffle($this->_deck);
        }
        public function printDeck()
        {
            foreach ($this->_deck as $card)
            {
                echo $card->toShortString()." | ";
            }
        }
        public function getCard()
        {
            return array_shift($this->_deck);
        }
    }


?>

File: python/deck.py
import logging
import random




class Deck:

    def __init__(self):
        self.deck = []
        for i in range(0, 4):
            for j in range(2, 15):
                self.deck.append(Card(j, i))

    def shuffle(self):
        random.shuffle(self.deck)

    @staticmethod
    def populate_byte_array():
        b = bytearray()
        for suit in range(0, 4):
            for rank in range(2, 15):
                b.append(rank<<2 | suit)
        return b

    def shuffle_perfect(self):
        # swap two cards to make suits meet
        temp = self.deck[13]
        self.deck[13] = self.deck[25]
        self.deck[25] = temp

        temp = self.deck[39]
        self.deck[39] = self.deck[51]
        self.deck[51] = temp

    def print_deck(self):
        for card in self.deck:
            logging.debug(card.to_string())

    def get_card(self):
        try:
            return self.deck.pop(0)
        except IndexError:
            return None

    def __str__(self):
        return ','.join(map(str, self.deck))


class SmallCard:
    def __init__(self, rank, suit):
        value = (rank << 2) | suit
        logging.info("rank: %d, suit: %d, value: %d" % (rank, suit, value))
        self.value = value

    def __str__(self):
        return str(self.value)

class SmallDeck:
    def __init__(self):
        self.deck = []
        for i in range(0, 4):
            for j in range(2, 15):
                self.deck.append((j<<2)|i)

class Card:

    rank_long = {
        2: 'Two', 3: 'Three', 4: 'Four', 5: 'Five', 6: 'Six',
        7: 'Seven', 8: 'Eight', 9: 'Nine', 10: 'Ten', 11: 'Jack',
        12: 'Queen', 13: 'King', 14: 'Ace'
    }
    rank_short = {
        2: '2', 3: '3', 4: '4', 5: '5', 6: '6',
        7: '7', 8: '8', 9: '9', 10: '0', 11: 'J',
        12: 'Q', 13: 'K', 14: 'A'
    }
    suit_long = {0: 'Spades', 1: 'Hearts', 2: 'Diamonds', 3: 'Clubs'}
    suit_short = {0: 's', 1: 'h', 2: 'd', 3: 'c'}

    def __init__(self, rank, suit):
        self.rank = rank
        self.suit = suit

    @staticmethod
    def small_card_to_string(sc):
        rank = sc >> 2
        suit = sc & 0b00000011
        return Card.rank_long[rank] + " of " + Card.suit_long[suit]

    def to_string(self):
        return Card.rank_long[self.rank] + " of " + Card.suit_long[self.suit]

    def to_short_string(self):
        return Card.rank_short[self.rank] + Card.suit_short[self.suit]

    def __str__(self):
        return Card.rank_short[self.rank] + Card.suit_short[self.suit]
